### ISSUE TITLE
What is the scope of no_std support?

### ISSUE THREAD

**Issue by yanshay** [shepmaster/snafu#467]
I'm trying to use Snafu in a no_std environment, and the basic example from the docs doesn't work.

It can't find `Whatever` and I checked the sourced and it's behind `std` feature.

So what is the scope of support for `no_std`?

**Comment by alerque** [comment]
As far as I understand it `whatever` itself does not support `no_std`, so this doesn't seem unexpected.

**Comment by yanshay** [comment]
> As far as I understand it `whatever` itself does not support `no_std`, so this doesn't seem unexpected.

Oh, any technical reason for that?
If I understand correctly `whatever` is the equivalent to `anyhow` crate and I recall I used it in the past in `no_std`.

**Comment by shepmaster** [comment]
Yes, `snafu::Whatever` requires the standard library for allocation (via `Box` and `String`) as well as the `Backtrace` type which should usually be provided by the standard library on modern Rust versions.

From anyhow's README (emphasis mine):

> To depend on Anyhow in no_std mode, disable our default enabled “std” feature in Cargo.toml. **A global allocator is required.**

So anyhow works with `libcore` + `liballoc`. I'd expect that SNAFU could do the same [^1] and just no one has needed that specific distinction yet. Do you have `liballoc` available?

[^1]: Or close to the same; you could always define your own `Whatever` at least.

**Comment by yanshay** [comment]
> So anyhow works with `libcore` + `liballoc`. I'd expect that SNAFU could do the same [1](#user-content-fn-1-85723774e9e455ef0fd8ade2e3db84b0) and just no one has needed that specific distinction yet. Do you have `liballoc` available?

I have `alloc`. In embedded there is a distinction between `std` and `alloc` since alloc is quite common to have.

**Comment by shepmaster** [comment]
> there is a distinction between `std` and `alloc`

I'm aware 😄. It's just that most of my embedded usage / knowledge of usage has been places with no allocation capability (e.g. tiny devices, Windows kernel drivers). In addition, I don't like using `Whatever` because I don't want to allocate if I can ever help it, plus I like having a known set of errors.

Anyway, can you try this branch and see if it works for your case? 

https://github.com/shepmaster/snafu/tree/alloc

**Comment by yanshay** [comment]
Great to see alloc feature added.
Does it mean that with no_atd alloc I can use all snafu capabilities or are there still things available only with std?

**Comment by shepmaster** [comment]
> use all snafu capabilities

There are still only things available with std. For example, returning [`Report`](https://docs.rs/snafu/latest/snafu/struct.Report.html) by returning it from `main` doesn't make sense if there's nowhere to return to.

That being said, if you find something that doesn't work with `core` + `alloc` that you expect to, let us know!
